controllers/imagehub/controllers/image_controller.go

```diff
@@ -61,66 +60,52 @@ type ImageReconciler struct {
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.13.0/pkg/reconcile
 func (r *ImageReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	r.Logger.V(1).Info("start reconcile for image")
-
-	// get image
 	img := &imagehubv1.Image{}
-	ctr := controller.Controller{
-		Client:   r.Client,
-		Logger:   r.Logger,
-		Eventer:  r.Recorder,
-		Operator: r,
-		Gvk: schema.GroupVersionKind{
-			Group:   imagehubv1.GroupVersion.Group,
-			Version: imagehubv1.GroupVersion.Version,
-			Kind:    "Image",
-		},
-		FinalizerName: imagehubv1.ImgFinalizerName,
+	if err := r.Get(ctx, req.NamespacedName, img); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
-	img.APIVersion = ctr.Gvk.GroupVersion().String()
-	img.Kind = ctr.Gvk.Kind
 
-	return ctr.Run(ctx, req, img)
-}
+	if ok, err := r.finalizer.RemoveFinalizer(ctx, img, r.doFinalizer); ok {
+		return ctrl.Result{}, err
+	}
 
-// Update .
-func (r *ImageReconciler) Update(ctx context.Context, req ctrl.Request, gvk schema.GroupVersionKind, obj client.Object) (ctrl.Result, error) {
-	r.Logger.V(1).Info("update reconcile controller image", "request", req)
-	img := &imagehubv1.Image{}
-	err := convert.JsonConvert(obj, img)
-	if err != nil {
-		r.Logger.V(2).Info("error in image json convert", "json", obj)
-		return ctrl.Result{Requeue: true}, err
+	if ok, err := r.finalizer.AddFinalizer(ctx, img); ok {
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		return r.doReconcile(ctx, img)
 	}
+	return ctrl.Result{}, errors.New("reconcile error from Finalizer")
+}
 
+func (r *ImageReconciler) doReconcile(ctx context.Context, obj client.Object) (ctrl.Result, error) {
+	r.Logger.V(1).Info("update reconcile controller image", "request", client.ObjectKeyFromObject(obj))
+	img, ok := obj.(*imagehubv1.Image)
+	if !ok {
+		return ctrl.Result{}, errors.New("obj convert Image is error")
+	}
 	pipelines := []func(ctx context.Context, img *imagehubv1.Image){
 		r.syncRepo,
 	}
 	for _, fn := range pipelines {
 		fn(ctx, img)
 	}
-
 	return ctrl.Result{}, r.Client.Update(ctx, img)
 }
 
-// Delete .
-func (r *ImageReconciler) Delete(ctx context.Context, req ctrl.Request, gvk schema.GroupVersionKind, obj client.Object) error {
-	r.Logger.V(1).Info("delete reconcile controller image", "request", req)
-	// get img from obj
-	img := &imagehubv1.Image{}
-	err := convert.JsonConvert(obj, img)
-	if err != nil {
-		r.Logger.V(2).Info("error in image json convert", "namespace/name", obj.GetNamespace()+"/"+obj.GetName())
-		return err
+func (r *ImageReconciler) doFinalizer(ctx context.Context, obj client.Object) error {
+	r.Logger.V(1).Info("delete reconcile controller image", "request", client.ObjectKeyFromObject(obj))
```

Comment:
this log will internally using json convert, this seems too heavy for just logs.

Reply:
how to solve it?

Reply:
https://github.com/labring/sealos/blob/70b4aaa70ca53dcb158f9a357c32aeaa05b4fb98/controllers/user/controllers/user_controller.go#L134